### PR TITLE
Apply modern CMake style

### PR DESCRIPTION
**Pull request by avaucher**
## Summary
To summarize this PR, consider the following example.
We want to compile [example1.cpp](https://github.com/rdkit/rdkit/blob/master/Docs/Book/C%2B%2BExamples/example1.cpp), with CMake. Currently, the minimal CMakeLists.txt for that would look like:

```
cmake_minimum_required( VERSION 2.8 )
project(ModernizingCMakeInRDKit)

find_package( rdkit REQUIRED )
find_package( Boost COMPONENTS graph system thread REQUIRED)

include_directories( ${RDKit_INCLUDE_DIRS})

add_executable(example1 example1.cpp)
target_link_libraries(example1 Boost::graph Boost::system Boost::thread)
target_link_libraries(example1 FileParsers_static SmilesParse_static RDGeometryLib_static RDGeneral_static GraphMol_static RDGeometryLib_static RDGeneral_static)
```

With modern CMake style (and if rdkit is compiled accordingly), this can be simplified to the following:

```
cmake_minimum_required( VERSION 3.1 )
project(ModernizingCMakeInRDKit)

find_package( rdkit REQUIRED )
find_package( Boost COMPONENTS graph system thread REQUIRED)

add_executable(example1 example1.cpp)
target_link_libraries(example1 Boost::graph Boost::system Boost::thread)
target_link_libraries(example1 FileParsers_static SmilesParse_static GraphMol_static)
```

Note the following differences:
* It isn't necessary anymore to set the include path
* The executable must only be linked against the libraries that are directly used in `example1.cpp`. Also, their order is not relevant.

To achieve such a simplification, the CMake code in `rdkit` must be changed, which is the object of this PR.

Note that to enable such functionality, CMake >= 3.1 is necessary.

## How this can be achieved
When setting up the CMake code for `rdkit` in a modern fashion, CMake is able to keep track of which libraries require which other libraries, which compile definitions, etc. When installing  `rdkit` with CMake, this information is then written automatically in the config files.
This means that if you want to use a high-level library `X` in another project, it is enough just to tell CMake that you want to use `X` without worrying about all its dependencies.

For more details about modern CMake, there is a good video [here](https://youtu.be/bsXLMQ6WgIk)

Switching to a more modern CMake syntax also avoids "contamining" the CMake code with raw `include_directories` and `link_libraries`, and makes the setup more modular and understandable.

## Possible extensions

The changes up to now are non-breaking (except the requirement for CMake 3.1), i.e. everything that used to work should continue working.

Further changes are possible, but they may be breaking: 
1. Use modern CMake syntax for dependencies. Many packages already use this new syntax. F.i.:
  * Boost
  * Qt
  Using the new syntax (i.e. just linking against `Qt5::Widgets` instead of `${Qt5Widgets_INCLUDE_DIRS}` and `${Qt5Widgets_LIBRARIES}`) allows to make the installation of `rdkit` independent of the exact Qt installation that was used for compiling `rdkit`.
2) Install CMake config files under standard location, i.e. `lib/cmake/rdkit`, instead of just `lib`
3) Add a namespace to the exported CMake targets.

I am happy to help in case of questions or for eventual extensions.

### CODE REVIEW

**Comment by bp-kelley** [comment]
@avaucher I like the idea behind this, note that it is breaking the swig builds.  

Please add the         
  
```
-DRDK_BUILD_SWIG_WRAPPERS=ON 
```

To the cmake flags for your internal test builds to see the issue.

File: Code/GraphMol/MolDraw2D/CMakeLists.txt [hunk 1]
```diff
@@ -12,45 +12,40 @@ rdkit_headers(MolDraw2D.h
...
-  include_directories(
-    ${Qt5Widgets_INCLUDE_DIRS}
-    ${Qt5OpenGL_INCLUDE_DIRS}
-    )
-
-  set(QTDRAW2D_SRC MolDraw2DQt.cpp)
-  set(EXTRA_LOCAL_LIBS ${EXTRA_LOCAL_LIBS} ${Qt5Widgets_LIBRARIES}  ${Qt5OpenGL_LIBRARIES} )
+  target_sources(MolDraw2D PRIVATE MolDraw2DQt.cpp)
+  target_include_directories(MolDraw2D PUBLIC ${Qt5Widgets_INCLUDE_DIRS} ${Qt5OpenGL_INCLUDE_DIRS})
+  target_compile_options(MolDraw2D PRIVATE "-std=c++11")
```
Comment (bp-kelley) [inline Code/GraphMol/MolDraw2D/CMakeLists.txt]: The c++ standard library can't be set here as this is a global setting and may actually be libstdc++

Comment (avaucher) [inline Code/GraphMol/MolDraw2D/CMakeLists.txt]: I added this line as a replacement of
`SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")`.
Should I delete it?

**Comment by greglandrum** [comment]
Given that cmake 3.x isn't available "out of the box" for many systems where people would want to build the rdkit, I would be a bit nervous about forcing v3.x.

However Switching to a more up-to-date style with cmake would be a great thing to do on the modern_c++ branch (to become the trunk at some point in the not-too-distant-future). 

@avaucher, if you're ok with the idea, I'd like to update the modern_cxx branch (https://github.com/rdkit/rdkit/tree/modern_cxx) and then switch the PR to be against that. I'll try and get the updates done today.

**Comment by avaucher** [comment]
@greglandrum : Sure, merging to modern_cxx is fine for me.

**Comment by greglandrum** [comment]
Great. Thanks. I'll let you know when I've managed to update that.

**Comment by greglandrum** [comment]
@avaucher : the modern_cxx branch is now up to day (in sync with master) and is a good target for this PR.

**Comment by avaucher** [comment]
@greglandrum I updated the target branch, but a few conflicts appeared; apparently the modern_cxx branch is not yet fully in sync with master?

**Comment by greglandrum** [comment]
Thanks for doing this. It's now merged onto the modern_cxx branch.

**Pull request merged**